include/seqan3/alphabet/mask/masked.hpp

```diff
@@ -100,39 +92,48 @@ class masked : public alphabet_tuple_base<masked<sequence_alphabet_t>, sequence_
     //!\brief Return a character.
     constexpr char_type to_char() const noexcept
     {
-        if (seqan3::to_rank(get<1>(*this)))
-        {
-            return to_lower(seqan3::to_char(get<0>(*this)));
-        }
-        else
-        {
-            return seqan3::to_char(get<0>(*this));
-        }
+        return rank_to_char[base_t::to_rank()];
     }
     //!\}
 
-    /*!\brief Validate whether a character value has a one-to-one mapping to an alphabet value.
-     *
-     * \details
-     *
-     * Satisfies the seqan3::semialphabet::char_is_valid_for() requirement via the seqan3::char_is_valid_for()
-     * wrapper.
-     *
-     * Default implementation: True for all character values that are reproduced by #to_char() after being assigned
-     * to the alphabet.
-     *
-     * ###Complexity
-     *
-     * Constant.
-     *
-     * ###Exceptions
-     *
-     * Guaranteed not to throw.
-     */
-    static constexpr bool char_is_valid(char_type const c) noexcept
+protected:
+    //!\brief Rank to char conversion table.
+    static constexpr std::array<char_type, alphabet_size> rank_to_char
     {
-        return masked{}.assign_char(c).to_char() == c;
-    }
+        [] ()
+        {
+            std::array<char_type, alphabet_size> ret{};
+
+            for (size_t i = 0; i < alphabet_size; ++i)
+            {
+                ret[i] = (i < alphabet_size / 2)
+                       ? seqan3::to_char(seqan3::assign_rank_to(i, sequence_alphabet_type{}))
```

Comment:
1. We should require that `sequence_alphabet_t` is default constructible then (in the template argument)
2. Why is the name `sequence_alphabet_t` and not just `alphabet_t`? Any special reason or just legacy code?

Reply:
> We should require that sequence_alphabet_t is default constructible then (in the template argument)

alphabet_tuple_base enforces this kind of (! reference, but it should be "semiregular").

Reply:
> Why is the name sequence_alphabet_t and not just alphabet_t? Any special reason or just legacy code?

There are two alphabets. The mask alphabet and the sequence_alphabet.